### PR TITLE
[CRIMAP-544] List uploaded evidence for in progress application

### PR DESCRIPTION
**Pull request by hiboabd**
## Description of change
This PR introduces the initial functionality to upload evidence for a draft application.

The `DocumentController` receives the file via the submit form and creates a local `Document` record with attributes like file_name etc. One attribute will be nil at this point, `object_key`.

The document is validated on size and content criteria and if valid, the `Upload` service requests an upload presigned URL via the datastore API and the presigned URL is then used to perform the upload directly to the bucket.

Upon successful completion of the upload, the returned `object_key` is written back to the `Document` record, and the transaction is finalised.

Upload failures are handled in the controller for JS and non-JS versions.

## Link to relevant ticket

[CRIMAP-544](https://dsdmoj.atlassian.net/browse/CRIMAP-544)

## Notes for reviewer
Apologies for the chunky PR!

Note: we have disabled multi file uploads for the non js version as the request payload included all the documents and it was not uploading one document at a time. As it's very unlikely that anyone will use the non js version, we did not feel it was worth the technical effort to handle both transaction types so the upload has been limited to one file at a time. 

Note: Application types list has been lifted from another application and may need dev review but should do for the time being. Design and content need reviewing, particularly for the error text.

Instructions to test are below.

### Before changes:
Screenshot 2023-09-12 at 17 36 54

### After changes:
Screenshot 2023-09-12 at 17 18 34

https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/c9c850a9-cdf8-4047-956a-fc9600b22e9d

## How to manually test the feature
This requires having both datastore and apply running

1. In your datastore repo, navigate to `main` and run `bundle install` to ensure gems are up to date. 
2. In one terminal, run `rails server` to spin up datastore 
3. In another terminal, run `docker-compose up localstack` to spin up localstack (required to upload to s3) 
4. In apply, navigate to this branch and run `bundle install` to ensure gems are up to date and spin up apply by running `rails server`. You may need to run a migration.
5. Then navigate to apply in your chosen browser and create a draft application (or use an existing draft application) and navigate to `/applications/{application_uuid}/steps/evidence/upload`. You should then be able to upload a file

Please disable JS in your chosen browser and test the non js upload functionality. 

[CRIMAP-544]: https://dsdmoj.atlassian.net/browse/CRIMAP-544?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/views/steps/evidence/upload/edit.en.html.erb [hunk 1]
```diff
@@ -7,32 +7,61 @@
...
+      <%= t('.upload_files_heading') %>
+      <span class="govuk-caption-m govuk-!-padding-top-2"><%= t('.upload_files_info') %></span>
+    </h2>
+
     <%= form_for @form_object, html: { id: 'dz-evidence-upload-form', class: 'dropzone' },
-                 url: crime_application_documents_path,
+                 url: bundle_documents_path(@form_object.current_document_bundle),
                  method: :post, multipart: true do |f| %>
 
+      <%= f.button t('.choose_files_button'),
```
Comment (arthurashman) [inline app/views/steps/evidence/upload/edit.en.html.erb:26]: When js is switched off, we've ended up with double the buttons, I think this whole button section needs deleting.
image

File: app/views/steps/evidence/upload/edit.en.html.erb [hunk 2]
```diff
@@ -7,32 +7,61 @@
...
+
       <div class="fallback">
         <%= f.govuk_file_field :documents, multiple: true,
                                label: { size: 'm' }, class: 'govuk-file-upload' %>
 
         <%= f.button t('.upload_file_button'),
                      class: 'govuk-button govuk-button--secondary',
                      data: { module: 'govuk-button' } %>
+
+        <div class="moj-multi-file__uploaded-files">
```
Comment (arthurashman) [inline app/views/steps/evidence/upload/edit.en.html.erb:40]: And the list is doubled up too, so this whole div needs deleting too (I think)

File: app/views/steps/evidence/upload/edit.en.html.erb (hunk 1 above)
Comment (arthurashman) [reply app/views/steps/evidence/upload/edit.en.html.erb:26]: Playing with it and switching js on and off, which 'Choose files' button shows needs to be conditional on if the fallback is used

Comment (arthurashman) [reply app/views/steps/evidence/upload/edit.en.html.erb:26]: I was looking into this and it's easier said than done I fear. I think we will need to give the js button a hidden class that is removed by js (therefore _not_ removed if js isn't on)

Comment (hiboabd) [reply app/views/steps/evidence/upload/edit.en.html.erb:26]: I've amended this

File: app/controllers/documents_controller.rb [hunk 3]
```diff
@@ -4,19 +4,58 @@ class DocumentsController < ApplicationController
...
+  end
+
+  def current_document_bundle
+    @current_document_bundle ||= DocumentBundle.find(document_bundle_id)
+  end
+
+  # TODO: unify if possible the submission params
+  # Currently XHR and HTML requests have different structure
+  # Also needs proper error handling for non-JS version if user click
+  # upload button without having selected a file
```
Comment (hiboabd) [inline app/controllers/documents_controller.rb:44]: Currently, this routes to the generic there is 'something wrong with the service' page. Can write a ticket to do this but i think this should go in to unblock other tickets

File: app/views/steps/evidence/upload/_uploaded_file.html.erb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<tr class="govuk-table__row">
+  <td id="<%=document.id%>" class="govuk-table__cell">
+    <span><%= document.filename %></span>
+
+    <% if document.s3_object_key %>
+      <strong class="govuk-tag govuk-tag--green">Uploaded</strong>
+    <% else %>
+      <% document.valid?(:storage) %>
```
Comment (hiboabd) [inline app/views/steps/evidence/upload/_uploaded_file.html.erb:8]: Error messages are not saved so when documents are retrieved they need to be validated against our criteria so the correct error message can be displayed

**Comment by arthurashman** [comment]
Just a tiny thing, but with JS on, the upload button remains with yellow focus state. It feels like it should return to grey after file selected

**Comment by arthurashman** [comment]
Hmmm v weird, that's how the example works - https://design-patterns.service.justice.gov.uk/components/multi-file-upload/

Although if you take the visually hidden and js to show the button out then drag and dropping doesn't have the continued highlight (although this _is_ on the MoJ design component.

Maybe ignore me and leave this as it is

File: app/validators/file_upload_validator.rb [hunk 5]
```diff
@@ -0,0 +1,57 @@
+class FileUploadValidator < ActiveModel::Validator
+  MIN_FILE_SIZE = 5 # KB
+  MAX_FILE_SIZE = 20 # MB
+
+  ALLOWED_CONTENT_TYPES = %w[
+    application/msword
+    application/pdf
+    application/rtf
+    application/vnd.ms-excel
```
Comment (arthurashman) [inline app/validators/file_upload_validator.rb:9]: Are we meant to be allowing excel files? It's not listed on the error on the design

Comment (arthurashman) [reply app/validators/file_upload_validator.rb:9]: Civil don't but do allow csv files

File: app/validators/file_upload_validator.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    application/vnd.oasis.opendocument.text
+    application/vnd.openxmlformats-officedocument.spreadsheetml.sheet
+    application/vnd.openxmlformats-officedocument.wordprocessingml.document
+    image/bmp
+    image/gif
+    image/jpeg
+    image/png
+    image/tiff
+    image/x-bitmap
+    text/csv
```
Comment (arthurashman) [inline app/validators/file_upload_validator.rb:19]: If we're going to allow these, we need to double check that's ok, and update the copy in the designs to say csv

File: app/validators/file_upload_validator.rb [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  MIN_FILE_SIZE = 5 # KB
+  MAX_FILE_SIZE = 20 # MB
+
+  ALLOWED_CONTENT_TYPES = %w[
+    application/msword
+    application/pdf
+    application/rtf
+    application/vnd.ms-excel
+    application/vnd.oasis.opendocument.text
+    application/vnd.openxmlformats-officedocument.spreadsheetml.sheet
```
Comment (arthurashman) [inline app/validators/file_upload_validator.rb:11]: This also not allowed according to the copy (and Civil don't allow it)

File: app/validators/file_upload_validator.rb [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    application/vnd.openxmlformats-officedocument.wordprocessingml.document
+    image/bmp
+    image/gif
```
Comment (arthurashman) [inline app/validators/file_upload_validator.rb:14]: This also not allowed according to the copy (and Civil don't allow it)

File: app/validators/file_upload_validator.rb (hunk 6 above)
Comment (arthurashman) [reply app/validators/file_upload_validator.rb:19]: Civil does allow (if that was the intention, to match what they do)

File: config/locales/en/steps.yml [hunk 9]
```diff
@@ -184,6 +184,14 @@ en:
           page_title: Upload supporting evidence
           heading: Upload supporting evidence
           upload_file_button: Upload file
+          upload_files_heading: Upload files
+          upload_files_info: The maximum file size is 20MB. Files must be a DOC, DOCX, RTF, ODT, JPG, BMP, PNG, TIF or PDF.
+          choose_files_button: Choose files
+          uploaded_files_heading: Uploaded files
+          uploaded_files_info: When you submit this application, the caseworker will also get these files.
+          file_name: File name
+          uploaded: Uploaded
```
Comment (arthurashman) [inline config/locales/en/steps.yml:193]: uploaded and uploading aren't used as far as I can tell

We could do with a translation for uploaded as it is hardcoded in `_uploaded_file.html.erb`

So:
```
file_name: File name
uploaded_file:
    uploaded: Uploaded
```

File: app/views/steps/evidence/upload/_uploaded_file.html.erb [hunk 10]
```diff
@@ -0,0 +1,15 @@
+<tr class="govuk-table__row">
+  <td id="<%=document.id%>" class="govuk-table__cell">
+    <span><%= document.filename %></span>
+
+    <% if document.s3_object_key %>
+      <strong class="govuk-tag govuk-tag--green">Uploaded</strong>
```
Comment (arthurashman) [inline app/views/steps/evidence/upload/_uploaded_file.html.erb:6]: If you add thee suggestion in the translation file, then we can do:
`      <strong class="govuk-tag govuk-tag--green"><%= t('.uploaded') %></strong>
`

**Review by arthurashman — APPROVED** [review]
This all looks great to me. 

Have obviously been using it lots in my own branch, so I'm also v happy that it all works as we want it to!

Nice work 😊

File: app/javascript/local/dropzone-cfg.js [hunk 11]
```diff
@@ -12,18 +12,74 @@ DropzoneCfg.prototype.init = function () {
...
+    this.$tableCell = document.getElementById(file.upload.uuid)
+    this.$tableCell.append(error)
+    this.$statusTag = this.$tableCell.querySelector(".govuk-tag")
+    this.$tableCell.removeChild(this.$statusTag)
+  })
 }
+function createTableRow(file) {
+  let row = document.createElement("tr")
+  let cell = document.createElement("td")
+  let statusTag = createStatusTag("Uploading")
```
Comment (arthurashman) [inline app/javascript/local/dropzone-cfg.js:52]: Yeah, this is hardcoded here, so I don't think you need it in the translation file

File: app/validators/file_upload_validator.rb (hunk 5 above)
Comment (hiboabd) [reply app/validators/file_upload_validator.rb:9]: This was lifted from civil and not reviewed 😄 so there's a mismatch between the what the designs list as acceptable file types and we what we have in the backend. I'll clean it the list to keep it in line with design

File: app/validators/file_upload_validator.rb (hunk 8 above)
Comment (hiboabd) [reply app/validators/file_upload_validator.rb:14]: I have no idea where this list came from because gifs is hilarious 😆

**Pull request merged**